cpp/src/fil/common.cuh

```diff
@@ -165,6 +166,103 @@ struct predict_params : shmem_size_params {
   int num_blocks;
 };
 
+namespace dispatch {
+
+template <template <bool, leaf_algo_t, int> class Func, typename storage_type,
+          bool cols_in_shmem, leaf_algo_t leaf_algo, int n_items,
+          typename... Args>
+void dispatch_final(predict_params& params, Args... args) {
+  Func<cols_in_shmem, leaf_algo, n_items>::template run<storage_type>(params,
+                                                                      args...);
+}
+
+template <template <bool, leaf_algo_t, int> class Func, typename storage_type,
```

Comment:
I wonder if there is a simpler solution not passing `template <...> class Func` every time.

In this case, I really don't know the answer to this question.